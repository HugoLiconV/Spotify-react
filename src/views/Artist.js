import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import GridContainer from '../components/Grid/GridContainer';
import GridItem from '../components/Grid/GridItem';
import { numberWithCommas } from '../services/utils';
import {
  getArtist,
  getArtistsAlbums,
  getArtistsTopTracks,
  getArtistsRelatedArtists
} from '../actions/artistActions';
import SongTable from '../components/SongTable';
import ImageGridList from '../components/Grid/ImageGridList';
import { withStyles } from '@material-ui/core/styles';
import { play } from '../actions/playerActions';
import { filterDataToDisplay, getWiderImage } from '../services/utils';
import uniqBy from 'lodash/uniqBy';

const styles = {
  coverImage: {
    backgroundPosition: '50%',
    backgroundRepeat: 'no-repeat',
    padding: '15px 0',
    backgroundSize: 'cover',
    margin: '0 auto 1em',
    width: '100%',
    height: '300px',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'flex-end',
    flexWrap: 'wrap'
  },
  text: {
    background: 'linear-gradient(60deg, #26c6da, #00acc1)',
    boxShadow:
      '0 12px 20px -10px rgba(0, 188, 212, 0.28), 0 4px 20px 0px rgba(0, 0, 0, 0.12), 0 7px 8px -5px rgba(0, 188, 212, 0.2)',
    borderRadius: '3px',
    color: 'white'
  },
  artistName: {
    padding: '0 10px'
  },
  followers: {
    padding: '1rem'
  }
};

export class Artist extends Component {
  componentDidMount() {
    const id = this.props.match.params.id;
    this.fetchArtistData(id);
  }

  componentDidUpdate(e) {
    if (e.history.location.pathname !== e.location.pathname) {
      const id = this.props.match.params.id;
      this.fetchArtistData(id);
    }
  }

  getArtist = id => {
    this.props.getArtist(id);
  };
  getArtistsAlbums = id => {
    this.props.getArtistsAlbums(id);
  };
  getArtistsTopTracks = id => {
    this.props.getArtistsTopTracks(id);
  };
  getArtistsRelatedArtists = id => {
    this.props.getArtistsRelatedArtists(id);
  };

  play = config => {
    this.props.play(config);
  };

  redirectToArtistInfo = id => {
    this.props.history.push(`${id}`);
  };

  fetchArtistData = id => {
    this.getArtist(id);
    this.getArtistsAlbums(id);
    this.getArtistsTopTracks(id);
    this.getArtistsRelatedArtists(id);
  };

  redirectToAlbumInfo = id => {
    this.props.history.push(`/album/${id}`);
  };

  onTileClick = ({ id, uri }) => {
    if (uri.includes('album')) {
      this.redirectToAlbumInfo(id);
    } else if (uri.includes('artist')) {
      this.redirectToArtistInfo(id);
    } else if (uri.includes('track')) {
      this.play({ uris: [uri] });
    } else if (uri.includes('playlist')) {
      alert('Not implemented yet ðŸ˜¢');
    }
  };

  render() {
    const {
      artist,
      artistsAlbums,
      artistsTopTracks,
      artistRelatedArtists,
      classes
    } = this.props;
    const artistImages = artist.images || [];
    const albumItems = artistsAlbums && artistsAlbums.items;
    const relatedArtistsItems =
      artistRelatedArtists && artistRelatedArtists.artists;
    const followers = (artist.followers && artist.followers.total) || 0;
    return (
      <div>
        <div
          style={{
            background: `url('${getWiderImage(artistImages)}')`
          }}
          className={classes.coverImage}
          alt={artist.name || ''}
        >
          <h2 className={classes.text + ' ' + classes.artistName}>
            {artist.name || ''}
          </h2>
          <h4 className={classes.text + ' ' + classes.followers}>
            Followers: {numberWithCommas(followers)}
          </h4>
        </div>
        <GridContainer justify="flex-start" alignItems="center">
          <GridItem xs={12} sm={12} md={6}>
            <SongTable
              songs={artistsTopTracks.slice(0, 5)}
              tableTitle="Top Tracks"
            />
          </GridItem>
          <GridItem xs={12} sm={12} md={6}>
            <h3>Related Artists</h3>
            <ImageGridList
              singleLine
              onTileClick={this.onTileClick}
              lg={3}
              xs={2}
              data={filterDataToDisplay(relatedArtistsItems, 300, 500)}
            />
          </GridItem>
          <GridItem xs={12} sm={12} md={12}>
            <h3>Albums</h3>
            <ImageGridList
              onTileClick={this.onTileClick}
              lg={4}
              xs={2}
              data={filterDataToDisplay(uniqBy(albumItems, 'name'))}
            />
          </GridItem>
        </GridContainer>
      </div>
    );
  }
}

Artist.propTypes = {
  artist: PropTypes.object.isRequired,
  artistsAlbums: PropTypes.object.isRequired,
  artistsTopTracks: PropTypes.array.isRequired,
  artistRelatedArtists: PropTypes.object.isRequired,
  getArtist: PropTypes.func.isRequired,
  getArtistsAlbums: PropTypes.func.isRequired,
  getArtistsTopTracks: PropTypes.func.isRequired,
  getArtistsRelatedArtists: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  const {
    artist,
    artistsAlbums,
    artistsTopTracks,
    artistRelatedArtists
  } = state.artist;
  return {
    artist,
    artistsAlbums,
    artistsTopTracks,
    artistRelatedArtists
  };
};

const mapDispatchToProps = {
  getArtist,
  getArtistsAlbums,
  getArtistsTopTracks,
  getArtistsRelatedArtists,
  play
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(Artist));
