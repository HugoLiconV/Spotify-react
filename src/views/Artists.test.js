import React from 'react';
import { shallow } from 'enzyme';
import { Artist } from './Artist';
import toJson from 'enzyme-to-json';

let props;
let wrapper;

beforeEach(() => {
  props = {
    match: { params: { id: 1 }, isExact: true, path: '', url: '' },
    artist: {
      name: 'Arctic Monkeys',
      images: [
        {
          height: 640,
          url: 'a.png',
          width: 640
        },
        {
          height: 320,
          url: 'b.png',
          width: 320
        },
        {
          height: 160,
          url: 'c.png',
          width: 160
        }
      ],
      followers: {
        href: null,
        total: 7216088
      }
    },
    artistsAlbums: {},
    artistsTopTracks: [],
    artistRelatedArtists: {},
    getArtist: jest.fn(),
    getArtistsAlbums: jest.fn(),
    getArtistsTopTracks: jest.fn(),
    getArtistsRelatedArtists: jest.fn(),
    classes: {
      coverImage: 'coverImage',
      artistName: 'artistName',
      followers: 'followers',
      text: 'text'
    }
  };
  wrapper = shallow(<Artist {...props} />);
});

describe('<Artist /> component', () => {
  it('should renders artist Info', () => {
    const title = wrapper.find('.artistName');
    const followers = wrapper.find('.followers');
    expect(title.text()).toBe('Arctic Monkeys');
    expect(followers.text()).toBe('Followers: 7,216,088');
  });

  it('should show cover with background image', () => {
    const cover = wrapper.find('.coverImage').get(0);
    // it shows the wider image, in this case is a.png
    expect(cover.props.style).toHaveProperty('background', "url('a.png')");
  });

  it('should match the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('getWiderImage method', () => {
  it('should return no_image_found when an empty array is passed', () => {
    const result = wrapper.instance().getWiderImage();
    expect(result).toBe('no_image_found.png');
  });

  it('should return no_image_found when undefined is passed', () => {
    const result = wrapper.instance().getWiderImage();
    expect(result).toBe('no_image_found.png');
  });

  it('should return wider image when passing an array of images', () => {
    const images = [
      {
        height: 640,
        url: 'a.png',
        width: 640
      },
      {
        height: 300,
        url: 'b.png',
        width: 300
      },
      {
        height: 64,
        url: 'c.png',
        width: 64
      }
    ];
    const result = wrapper.instance().getWiderImage(images);
    expect(result).toBe('a.png');
  });
});

describe('getItemImage method', () => {
  it('should return no_image_found when passing undefined', () => {
    const result = wrapper.instance().getItemImage();
    expect(result).toBe('no_image_found.png');
  });

  it('should return no_image_found when passing an emtpy array', () => {
    const result = wrapper.instance().getItemImage([]);
    expect(result).toBe('no_image_found.png');
  });

  it('should return image with expected size', () => {
    const images = [
      {
        height: 640,
        url: 'a.png',
        width: 640
      },
      {
        height: 300,
        url: 'b.png',
        width: 300
      },
      {
        height: 64,
        url: 'c.png',
        width: 64
      }
    ];
    const result = wrapper.instance().getItemImage(images, 200, 350);
    expect(result).toBe('b.png');
  });

  it('should return first image from array when doesnt find one with expected size', () => {
    const images = [
      {
        height: 640,
        url: 'a.png',
        width: 640
      },
      {
        height: 300,
        url: 'b.png',
        width: 300
      },
      {
        height: 64,
        url: 'c.png',
        width: 64
      }
    ];
    const result = wrapper.instance().getItemImage(images, 700, 800);
    expect(result).toBe('a.png');
  });
});
