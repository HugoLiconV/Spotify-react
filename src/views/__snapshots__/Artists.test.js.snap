// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Artist /> component should match the snapshot 1`] = `
<div>
  <div
    alt="Arctic Monkeys"
    className="coverImage"
    style={
      Object {
        "background": "url('a.png')",
      }
    }
  >
    <h1
      className="text artistName"
    >
      Arctic Monkeys
    </h1>
    <h4
      className="text followers"
    >
      Followers: 
      7,216,088
    </h4>
  </div>
  <GridContainer
    alignItems="center"
    justify="flex-start"
  >
    <WithStyles(GridItem)
      md={6}
      sm={12}
      xs={12}
    >
      <WithStyles(SongTable)
        songs={Array []}
        tableTitle="Top Tracks"
      />
    </WithStyles(GridItem)>
    <WithStyles(GridItem)
      md={6}
      sm={12}
      xs={12}
    >
      <h3>
        Related Artists
      </h3>
      <WithStyles(WithTheme(WithWidth(SingleLineGridList)))
        data={Array []}
        lg={3}
      />
    </WithStyles(GridItem)>
    <WithStyles(GridItem)
      md={12}
      sm={12}
      xs={12}
    >
      <h3>
        Albums
      </h3>
      <WithStyles(WithTheme(WithWidth(SingleLineGridList)))
        data={Array []}
        grid={true}
        lg={4}
      />
    </WithStyles(GridItem)>
  </GridContainer>
</div>
`;
